imapclient/imapclient.py

```diff
@@ -761,6 +768,26 @@ def idle(self):
         if resp is not None:
             raise exceptions.IMAPClientError('Unexpected IDLE response: %s' % resp)
 
+    def _poll_socket(self, sock, timeout=None):
+        """
+        Polls the socket for events telling us it's available to read.
+        This implementation is more scalable because it ALLOWS your process
+        to have more than 1024 file descriptors.
+        """
+        poller = select.poll()
+        poller.register(sock.fileno(), select.POLLIN)
+        timeout = timeout * 1000 if timeout is not None else None
+        return poller.poll(timeout)
+
+    def _select_poll_socket(self, sock, timeout=None):
+        """
+        Polls the socket for events telling us it's available to read.
+        This implementation is a fallback because it FAILS if your process
+        has more than 1024 file descriptors.
+        We still need this for Windows and some other niche systems.
+        """
+        return select.select([sock], [], [], timeout)[0]
```

Comment:
Would you mind moving `_poll_socket` and `_select_poll_socket` out to standalone funcs in a new module (`poll.py`?). This module could then have a `get_poll_func` function which returns the appropriate variant depending on what's available. 

This module should have it's own unit tests.

This would be a little cleaner and will probably make the tests simpler too (just patch `get_poll_func`).